src/main/java/io/cdap/plugin/google/sheets/sink/GoogleSheetsRecordWriter.java

```
@@ -0,0 +1,373 @@
+    private final String spreadSheetName;
+
+    private RecordsBatch(List<FlatternedRowsRequest> group, String spreadSheetName) {
+      this.group = group;
+      this.spreadSheetName = spreadSheetName;
+    }
+
+    public void add(FlatternedRowsRequest record) {
+      group.add(record);
+    }
+
+    public List<FlatternedRowsRequest> getGroup() {
+      return group;
+    }
+
+    public String getSpreadSheetName() {
+      return spreadSheetName;
+    }
+
+    @Override
+    public int compareTo(RecordsBatch o) {
+      return o.group.size() - this.group.size();
+    }
+  }
+
+  /**
+   * Task that executes all requests from {@link RecordsBatch} by single Sheets API request.
+   */
+  private class MessagesProcessor implements Callable {
+
+    private final RecordsBatch recordsBatch;
+
+    private MessagesProcessor(RecordsBatch recordsBatch) {
+      this.recordsBatch = recordsBatch;
+    }
+
+    @Override
+    public Object call() throws Exception {
+      try {
+        List<FlatternedRowsRequest> recordsToProcess = recordsBatch.getGroup();
+        if (recordsToProcess.isEmpty()) {
+          return null;
+        }
+        String spreadSheetName = recordsBatch.getSpreadSheetName();
+
+        List<Request> contentRequests = new ArrayList<>();
+        List<Request> mergeRequests = new ArrayList<>();
+        List<String> sheetTitles = new ArrayList<>();
+        for (FlatternedRowsRequest flatternedRowsRequest : recordsToProcess) {
+          contentRequests.add(flatternedRowsRequest.getContentRequest());
+          mergeRequests.addAll(flatternedRowsRequest.getMergeRequests());
+          sheetTitles.add(flatternedRowsRequest.getSheetTitle());
+        }
+
+        String spreadSheetId = availableFiles.get(spreadSheetName);
+
+        sheetsSinkClient.populateCells(spreadSheetId, spreadSheetName, sheetTitles,
+          contentRequests, mergeRequests);
+      } catch (Exception e) {
+        LOG.error(e.toString());
```

Don't ignore exceptions.

---

removed